README.md
# OpenThaiGPT

[![](https://img.shields.io/pypi/v/openthaigpt.svg)](https://pypi.python.org/pypi/openthaigpt) [![](https://pyup.io/repos/github/OpenThaiGPT/openthaigpt/shield.svg)](https://pyup.io/repos/github/OpenThaiGPT/openthaigpt/)

OpenThaiGPT focuses on developing a Thai Chatbot system to have capabilities equivalent to ChatGPT, as well as being able to connect to external systems and be able to retrieve data flexibly. Easily expandable and customizable and developed into Free open source software for everyone.

* Free software: Apache Software License 2.0
* Project Homepage: https://openthaigpt.aieat.or.th
* Documentation: https://openthaigpt.readthedocs.io

## Versions

- OpenThaiGPT chat inference library (this repository): version 0.1.1

- Released models
    * kobkrit/openthaigpt-0.1.0-beta
      - Pretraining Model: Facebook Llama (7 billion params)
      - Dataset: 200,000 Various Translated Instruction Dataset 
      - RLHF: None
      - Minimium Requirement: Nvidia T4 16GB

    * kobkrit/openthaigpt-0.1.0-alpha
      - Pretraining Model: ByT5-XL (3.74 billion params)
      - Dataset: 50,000 Thai SelfInstruct 
      - RLHF: None
      - Minimium Requirement: Nvidia A100 40GB

    * kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.4
      - Pretraining Model: GPT-2 Thai-base
      - InstructDataset: 300,000 Pantip + 5,000 Wiki QA => 12,920 Thai InstructGPT
      - RLHF: None
      - Minimium Requirement: CPU or Nvidia GTX 1060 6GB

    * kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.3
      - Pretraining Model: GPT-2 Thai-base
      - InstructDataset: 300,000 Pantip + 5,000 Wiki QA => 7,000 Thai InstructGPT
      - RLHF: None
      - Minimium Requirement: CPU or Nvidia GTX 1060 6GB

    * kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.2
      - Pretraining Model: GPT-2 Thai-base
      - InstructDataset: 7,000 Thai InstructGPT
      - RLHF: None
      - Minimium Requirement: CPU or Nvidia GTX 1060 6GB

    * kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.1
      - Pretraining Model: GPT-2 Thai-base
      - InstructDataset: 298,678 QA Pairs getting from 70,000 Pantip katoos + Wikipedia QA by iApp
      - RLHF: None
      - Minimium Requirement: CPU or Nvidia GTX 1060 6GB


## Installation
Python>=3.6

### CPU-Only
``$ pip install openthaigpt torch --extra-index-url https://download.pytorch.org/whl/cpu``

### With GPU

CUDA 11.6
``$ pip install openthaigpt torch --extra-index-url https://download.pytorch.org/whl/cu116``

CUDA 11.7
``$ pip install openthaigpt torch``

## Using 0.1.0-beta model
```
import openthaigpt

print(openthaigpt.generate(instruction="แปลภาษาอังกฤษเป็นภาษาไทย", input="We want to reduce weight.", model_name = "kobkrit/openthaigpt-0.1.0-beta", min_length=50, max_length=300, top_p=0.75, top_k=40, num_beams=1, no_repeat_ngram_size=0, temperature=0.1, early_stopping=True, load_8bit=False))
```

## Using 0.1.0-alpha model
```
import openthaigpt

print(openthaigpt.generate(instruction="แปลภาษาอังกฤษเป็นภาษาไทย", input="We want to reduce weight.", model_name = "kobkrit/openthaigpt-0.1.0-alpha", min_length=50, max_length=300,  top_k=20, num_beams=5, no_repeat_ngram_size=20, temperature=1, early_stopping=True))
```


## Usage 0.0.1-0.0.4 model
```
import openthaigpt

print(openthaigpt.generate("Q: อยากลดความอ้วนทำไง\n\nA:", model_name = "kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.4"))
print(openthaigpt.zero("การลดน้ำหนักเป็นเรื่องที่ต้องพิจารณาอย่างละเอียดและรอบคอบเพื่อให้ได้ผลลัพธ์ที่ดีและมีประสิทธิภาพมากที่สุด"))
```

## Sponsored by
* Pantip.com
* ThaiSC

## Collaborated By
* Artificial Intelligence Entrepreneur Association of Thailand (AIEAT)
* Artificial Intelligence Association of Thailand (AIAT)

## Supported By
* NECTEC
* iApp Technology
* NVIDIA
* Microsoft
* Mahidol University
* Gitbook


HISTORY.md
# History

0.1.1 (2023-05-26)
* Add Model version 0.1.0-beta (https://huggingface.co/kobkrit/openthaigpt-0.1.0-beta)

0.1.0 (2023-04-23)
* Support Model Selection
* Add Model version 0.1.0-alpha (https://huggingface.co/kobkrit/openthaigpt-0.1.0-alpha) to supported models.

0.0.9 - 0.0.10 (2023-03-19)

* Release OpenThaiGPT Zero

0.0.8 (2023-03-12)

* Update Readme

0.0.7 (2023-03-12)

* Update Model to Version 0.0.4 (https://huggingface.co/kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.4)

0.0.6 (2023-03-05)

* Update README

0.0.5 (2023-02-28)

* First release on PyPI.


tests/__init__.py
"""Unit test package for openthaigpt."""


tests/test_openthaigpt.py
#!/usr/bin/env python

"""Tests for `openthaigpt` package."""


import unittest

from openthaigpt import openthaigpt_module

class TestOpenthaigpt(unittest.TestCase):
    """Tests for `openthaigpt` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_0000_usage(self):
        """Test something."""
        answer = openthaigpt_module.generate("Q: สวัสดีครับ")
        print(answer)
        assert answer is not None


openthaigpt/openthaigpt.py
"""Main module."""
from transformers import GPT2LMHeadModel, GPT2Tokenizer
import pandas as pd
import torch
from torch.utils.data import Dataset, random_split
from transformers import GPT2Tokenizer, TrainingArguments, Trainer, GPT2LMHeadModel

pretrained_name = "kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.2"

tokenizer = GPT2Tokenizer.from_pretrained(pretrained_name, bos_token='<|startoftext|>',unk_token='<|unk|>', eos_token='<|endoftext|>', pad_token='<|pad|>')
model = GPT2LMHeadModel.from_pretrained(pretrained_name).cuda()
model.resize_token_embeddings(len(tokenizer))

def generate(input, max_length=300, top_k=50, top_p=0.95, num_beam=5, no_repeat_ngram_size=2, early_stopping=True, temperature=1.9):
    generated = tokenizer("<|startoftext|>"+input, return_tensors="pt").input_ids.cuda()
    output = model.generate(generated, top_k=top_k, num_beams=num_beam, no_repeat_ngram_size=no_repeat_ngram_size, 
        early_stopping=early_stopping, max_length=max_length, top_p=top_p, temperature=temperature)
    return tokenizer.decode(output[0], skip_special_tokens=True)

openthaigpt/openthaigpt_module.py
"""Main module."""
from transformers import GPT2Tokenizer, GPT2LMHeadModel
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer, AutoModelForCausalLM

from transformers.models.mt5 import MT5Tokenizer
from transformers import DataCollatorForLanguageModeling
from peft import PeftModel, PeftConfig
from transformers import GenerationConfig, LlamaForCausalLM, LlamaTokenizer
from evaluate import load
import torch
import sys

# Check if CUDA is available
if torch.cuda.is_available():
    # "to(dev)" equals "cuda()" if CUDA is available.
    dev = torch.device('cuda')
else:
    dev = torch.device('cpu')

pretrained_name = "kobkrit/openthaigpt-0.1.0-beta"
tokenizer = None
model = None

def generate(input, instruction="", model_name = "kobkrit/openthaigpt-0.1.0-beta", min_length=0, max_length=256, top_k=40, top_p=0.75, num_beams=1, no_repeat_ngram_size=0, early_stopping=True, temperature=0.1, load_8bit=False):
    global tokenizer, model
    # load model
    if (not tokenizer or not model):
        if model_name == "kobkrit/openthaigpt-0.1.0-beta":
            base_model = 'decapoda-research/llama-7b-hf'
            lora_weights = 'kobkrit/openthaigpt-0.1.0-beta'

            tokenizer = LlamaTokenizer.from_pretrained(base_model)
            model = LlamaForCausalLM.from_pretrained(
                    base_model,
                    load_in_8bit=load_8bit,
                    torch_dtype=torch.float16,
                    device_map="auto",
                )
            model = PeftModel.from_pretrained(
                model,
                lora_weights,
                torch_dtype=torch.float16,
            ).to(dev)

            # unwind broken decapoda-research config
            model.config.pad_token_id = tokenizer.pad_token_id = 0  # unk
            model.config.bos_token_id = 1
            model.config.eos_token_id = 2

            if not load_8bit:
                model.half()  # seems to fix bugs for some users.

            # model.eval()
            if torch.__version__ >= "2" and sys.platform != "win32":
                model = torch.compile(model)

        elif model_name == "kobkrit/openthaigpt-0.1.0-alpha":
            config = PeftConfig.from_pretrained(model_name)
            model = AutoModelForSeq2SeqLM.from_pretrained(config.base_model_name_or_path)
            model = PeftModel.from_pretrained(model, model_name).to(dev)
            tokenizer = AutoTokenizer.from_pretrained(config.base_model_name_or_path)
        else:
            tokenizer = GPT2Tokenizer.from_pretrained(model_name, bos_token='<|startoftext|>',unk_token='<|unk|>', eos_token='<|endoftext|>', pad_token='<|pad|>')
            model = GPT2LMHeadModel.from_pretrained(model_name).to(dev)

    # inference
    if model_name == "kobkrit/openthaigpt-0.1.0-beta":
        if (input and not instruction):
          instruction = input
          input = ""
        
        prompt = f"Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\n{instruction}\n\n### Response:\n"
        if (instruction and input):
          prompt = f"Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:\n"
  
        inputs = tokenizer(prompt, return_tensors="pt")
        input_ids = inputs["input_ids"].to(dev)
        generation_config = GenerationConfig(
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            num_beams=num_beams,
            early_stopping=early_stopping,
            min_new_tokens=min_length,
            max_new_tokens=max_length,
            no_repeat_ngram_size=no_repeat_ngram_size
        )
        # Without streaming
        with torch.no_grad():
            generation_output = model.generate(
                input_ids=input_ids,
                generation_config=generation_config,
                return_dict_in_generate=True,
                output_scores=True
            )
        s = generation_output.sequences[0]
        output = tokenizer.decode(s).split("### Response:")[1].strip()
        return output

    elif model_name == "kobkrit/openthaigpt-0.1.0-alpha":
        generated = tokenizer('<instruction>: ' + str(instruction) + ' <input>: ' + str(input) + ' <output>: ', max_length=max_length, padding="max_length", truncation=True, return_tensors="pt").input_ids.to(dev)

        with torch.no_grad():
            output = model.generate(input_ids=generated, top_k=top_k, num_beams=num_beams, no_repeat_ngram_size=no_repeat_ngram_size, 
                early_stopping=early_stopping, min_length=min_length, max_length=max_length, top_p=top_p, temperature=temperature)
            return tokenizer.decode(output[0], skip_special_tokens=True)
    else:
        generated = tokenizer("<|startoftext|>"+input, return_tensors="pt").input_ids.to(dev)
        
        with torch.no_grad():
            output = model.generate(input_ids=generated, top_k=top_k, num_beams=num_beams, no_repeat_ngram_size=no_repeat_ngram_size, 
                early_stopping=early_stopping, min_length=min_length, max_length=max_length, top_p=top_p, temperature=temperature)
            return tokenizer.decode(output[0], skip_special_tokens=True)


def zero(input, model_name="kobkrit/openthaigpt-gpt2-instructgpt-poc-0.0.4", threshold=10):
    perplexity = load("perplexity", module_type="metric")
    results = perplexity.compute(predictions=[input], model_id=model_name)
    score = results['perplexities'][0]
    isGeneratedFromOpenThaiGPT = False
    if (score < threshold):
        isGeneratedFromOpenThaiGPT = True
    return {'perplexity':score, 'threshold':threshold, 'isGeneratedFromOpenThaiGPT':isGeneratedFromOpenThaiGPT}
